protowhat/sct_syntax.py

```diff
@@ -1,134 +1,329 @@
-        self._stack = [] if stack is None else stack
-
-    def __call__(self, *args, **kwargs):
-        if self._crnt_sct:
-            # calling an SCT function (after attribute access)
-            call_data = (self._crnt_sct, args, kwargs)
-            return self.__class__(self._stack + [call_data], self._attr_scts)
+            self._state = state
+
+    def __str__(self):
+        if self.call is None:
+            result = "Ex()"
         else:
-            # running the chain
-            state = kwargs.get("state") or args[0]
-            return reduce(
-                lambda s, cd: self._call_from_data(s, *cd), self._stack, state
+            result = str(self.call)
+
+        if len(result) and len(self.next) > 0:
+            result += "."
+
+        if len(self.next) == 1:
+            result += str(self.next[0])
+        elif len(self.next) > 1:
+            result += "multi({})".format(", ".join(map(str, self.next)))
+
+        return result
+
+
+class ChainExtender:
+    """
+    This handles branching off (multiple next)
+    either from a Chain or a chain attribute
+    """
+
+    def __init__(self, chain: Chain, function: Callable):
+        self.chain = chain
+        self.function = function
+
+    def __call__(self, *args, **kwargs) -> Chain:
+        return type(self.chain)(
+            ChainedCall(self.function, args, kwargs), previous=self.chain
+        )
+
+    def __getattr__(self, item):
+        self.invalid_next_step(item)
+
+    def __rshift__(self, other):
+        self.invalid_next_step(other)
+
+    def invalid_next_step(self, next_step: str):
+        raise AttributeError(
+            "Expected a call of {} before {}. ".format(
+                getattr(self.function, "__name__", repr(self.function)), next_step
             )
+        )
+
+
+def get_chain_ends(chain: Chain) -> List[Chain]:
+    if chain.next:
+        return [*chain_iters(*(get_chain_ends(branch) for branch in chain.next))]
+    else:
+        return [chain]
+
+
+class ChainStart:
+    """Create new chains and keep track of the created chains"""
+
+    def __init__(self, chainable_functions: Dict[str, Callable]):
+        self.chain_roots = []
+        self.chainable_functions = chainable_functions
 
-    @staticmethod
-    def _call_from_data(state, f, args, kwargs):
-        return link_to_state(f)(state, *args, **kwargs)
+    def __call__(self) -> Chain:
+        """Create a new chains and store it"""
+        raise NotImplementedError()
 
-    @classmethod
-    def _from_func(cls, f, *args, _attr_scts=None, **kwargs):
-        """Creates a function chain starting with the specified SCT (f), and its arguments."""
-        func_chain = cls(attr_scts=_attr_scts)
-        func_chain._stack.append([f, args, kwargs])
-        return func_chain
+    def register_chainable_function(self, function: Callable, name: str = None):
+        name = name if name is not None else function.__name__
+        self.chainable_functions[name] = function
```

Comment:
This is a mutation and could be confusing (as discussed in PM)

Reply:
I will document it here and in `__init__`.